operations/ftdc.go

```diff
@@ -737,19 +737,86 @@ func toT2() cli.Command {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
 
-			inputFile, err := os.Open(inputPath)
+			inputStat, err := os.Stat(inputPath)
 			if err != nil {
 				return errors.Wrapf(err, "problem opening file '%s'", inputPath)
 			}
-			defer func() { grip.Warning(inputFile.Close()) }()
 
-			// All genny output files are named using the workload actor and operation.
-			//   i.e., Actor.Operation.ftdc
-			//
-			// We get the actor and operation names from the ftdc filepath for use
-			// in the translation process.
-			fileName := filepath.Base(inputPath)
-			actorOperation := strings.Split(fileName, ".ftdc")
+			// Stores ftdc chunk iterators and its metadata to be used in TranslateGenny.
+			var outputSlice []*ftdc.GennyOutputMetadata
+
+			// Check if input path is an individual file or directory.
+			// All ftdc files in the provided input directory are initially
+			// used to set GennyOutputMetadata Name, Iter, StartTime and EndTime.
+			switch mode := inputStat.Mode(); {
+			// Handle directory.
+			case mode.IsDir():
+				if !strings.HasSuffix(inputPath, "/") {
+					inputPath += "/"
+				}
+				input, err := os.Open(inputPath)
+				if err != nil {
+					return errors.Wrapf(err, "problem opening input path '%s'", inputPath)
+				}
+				defer func() { grip.Warning((input.Close())) }()
+
+				files, err := input.Readdir(-1)
+				if err != nil {
+					return errors.Wrapf(err, "problem reading dir '%s'", inputPath)
+				}
+
+				for _, file := range files {
+					if file.Mode().IsRegular() && filepath.Ext(file.Name()) == ".ftdc" {
+						var gennyOutput ftdc.GennyOutputMetadata
+						filePath := filepath.Join(inputPath + file.Name())
+
+						f, err := os.Open(filePath)
+						if err != nil {
+							return errors.Wrapf(err, "problem opening file '%s'", filePath)
+						}
+						defer func() { grip.Warning(f.Close()) }()
+						gennyOutput.Iter = ftdc.ReadChunks(ctx, f)
+
+						// GetGennyTime exhausts the current chunk iterator and renders it
+						// unusable for future tasks.
+						gennyOutput = ftdc.GetGennyTime(ctx, gennyOutput)
+
+						// Reopen the file to get a new chunk iterator for the ftdc file.
+						f, err = os.Open(filePath)
+						if err != nil {
+							return errors.Wrapf(err, "problem opening file '%s'", filePath)
+						}
+
+						gennyOutput.Iter = ftdc.ReadChunks(ctx, f)
+						gennyOutput.Name = strings.Split(file.Name(), ".ftdc")[0]
+						// Upon describing the genny output metadata, store into the slice
+						// to be used in TranslateGenny.
+						outputSlice = append(outputSlice, &gennyOutput)
+					}
+				}
+			// Handle single file.
+			case mode.IsRegular():
+				var gennyOutput ftdc.GennyOutputMetadata
+				input, err := os.Open(inputPath)
+				if err != nil {
+					return errors.Wrapf(err, "problem opening file '%s'", inputPath)
+				}
+				defer func() { grip.Warning((input.Close())) }()
```

Comment:
Since you reuse the variable `input` on L809, this defer will only close the second instance of that variable. I would just call `input.Close` on L808 without the defer since you don't really need it. And then again on L817. Defering closes are only useful if there are a lot of error cases (so you don't need to close on every error case).

Reply:
Thanks, that makes sense. The defer statement would probably close on the last file that gets read. 

I couldn't call input.Close on L817 because TranslateGenny seems to rely on the files remaining open otherwise this error throws:
`runtime error: invalid memory address or nil pointer dereference`